### PR TITLE
Improve static type checking

### PR DESCRIPTION
**Pull request by dklimpel**
Signed-off-by: Dirk Klimpel [dirk@klimpel.org](mailto:dirk@klimpel.org)

### CODE REVIEW

**Comment by richvdh** [comment]
> I am not sure if this error is realted to this PR:

It's happening on my PR too (#334), so I guess not

**Comment by dklimpel** [comment]
I have seen that the pipeline also uses Python 3.7. It is EOL in 2 month.

File: tests/test_gcm.py [hunk 1]
```diff
@@ -57,30 +61,34 @@ class TestGcmPushkin(GcmPushkin):
...
+    def __init__(self, name: str, sygnal: "Sygnal", config: Dict[str, Any]):
         super().__init__(name, sygnal, config)
-        self.preloaded_response = None
-        self.preloaded_response_payload = None
-        self.last_request_body = None
-        self.last_request_headers = None
+        self.preloaded_response = DummyResponse(0)
+        self.preloaded_response_payload: Dict[str, Any] = {}
+        self.last_request_body: Dict[str, Any] = {}
+        self.last_request_headers: Dict[AnyStr, List[AnyStr]] = {}  # type: ignore[valid-type]
```
Comment (DMRobertson) [inline tests/test_gcm.py:69]: What's the complaint here?

File: tests/test_apns.py
Comment (DMRobertson) [inline tests/test_apns.py:1]: Is it now possible to mark these files as `disallow_untyped_defs = true` in mypy.ini?

**Review by DMRobertson — COMMENTED** [review]
Thanks for this. Just a couple of questions:

File: tests/test_gcm.py (hunk 1 above)
Comment (dklimpel) [reply tests/test_gcm.py:69]: `AnyStr` ist not allowed there. But `last_request_headers` has different types. In one place `str` in another `bytes`.

Comment (dklimpel) [reply tests/test_gcm.py:69]:
```
tests/test_gcm.py:69: error: Type variable "typing.AnyStr" is unbound  [valid-type]
tests/test_gcm.py:69: note: (Hint: Use "Generic[AnyStr]" or "Protocol[AnyStr]" base class to bind "AnyStr" inside a class)
tests/test_gcm.py:69: note: (Hint: Use "AnyStr" in function signature to bind "AnyStr" inside a function)
```

Comment (DMRobertson) [reply tests/test_gcm.py:69]: Hmm. Does the following patch introduce any new errors?

```diff
diff --git a/tests/test_gcm.py b/tests/test_gcm.py
index 3a4f61f..589165c 100644
--- a/tests/test_gcm.py
+++ b/tests/test_gcm.py
@@ -13,7 +13,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import json
-from typing import TYPE_CHECKING, Any, AnyStr, Dict, List, Tuple
+from typing import TYPE_CHECKING, Any, AnyStr, Dict, List, Tuple, Union
 
 from sygnal.gcmpushkin import GcmPushkin
 
@@ -66,7 +66,7 @@ def __init__(self, name: str, sygnal: "Sygnal", config: Dict[str, Any]):
         self.preloaded_response = DummyResponse(0)
         self.preloaded_response_payload: Dict[str, Any] = {}
         self.last_request_body: Dict[str, Any] = {}
-        self.last_request_headers: Dict[AnyStr, List[AnyStr]] = {}  # type: ignore[valid-type]
+        self.last_request_headers: Union[Dict[str, List[str]], Dict[bytes, List[bytes]]] = {}
         self.num_requests = 0
 
     def preload_with_response(
```

File: tests/test_apns.py
Comment (dklimpel) [reply tests/test_apns.py:1]: Done

File: tests/test_gcm.py (hunk 1 above)
Comment (dklimpel) [reply tests/test_gcm.py:69]: Yes,
`tests/test_gcm.py:69: error: Incompatible types in assignment (expression has type "Dict[<nothing>, <nothing>]", variable has type "Union[Dict[str, List[str]], Dict[bytes, List[bytes]]]")  [assignment]`

Comment (DMRobertson) [reply tests/test_gcm.py:69]: Okay, let's leave this as it you propose for now.

**Pull request merged**